### PR TITLE
HUD Vision Datums

### PR DESCRIPTION
**Pull request by jwhitak**
# HUD Vision Datums
Datums are good, right? This is a back end update that moves most of the hardcoded HUDs/scanners (security, medical, diagnostic, mesons, night vision, etc) into generic datums that can be utilized on objects/mobs to apply their effects. It also adds a few new glasses and effects (unused in normal station play) for testing/bussing purposes. Average player should see no differences before and after this PR.

Things should be at feature-parity, so you should notice no differences in their use. If you do notice any oddities, plrease make a bug report.

Coders, you can use them by calling ``apply_hud(instance)`` or ``apply_hud_by_type(/datum/visioneffect/yourtypehere)`` on your items. You can also ``remove_hud`` in the same ways. glasses/hud apply all their visioneffects at once, glasses/scanners toggle them all on or off at once.

Admins, you can troll your friends by right clicking them, proccalling, ``apply_hud_by_type``, arguments 1, datum ``/datum/visioneffect/youreffecthere``. Removing it is the same proc, but with remove instead of apply. While you can manually manipulate the hud list in their mob variables, it's strongly recommended to use the procs because manually adding/removing the datum instances will not call on_apply and on_remove for the huds, which will break some of them. Notably the scanners-style ones like mesons.

To bus your own custom glasses with huds of your choice, currently you can only use a ``/obj/clothing/glasses/hud``, ``/obj/clothing/glasses/scanner``, or any subtype of those. To add a new hud, use the proc ``add_new_hud_by_type``, arguments 1, then ``/datum/visioneffect/youreffecthere``. To remove existing huds, you have to vv the item and remove the relevant hud from both ``hud_types`` and ``stored_huds``.

The current options:
```
/datum/visioneffect/accountdb/wage
// Shows mob's ID's current wage in the bottom left
/datum/visioneffect/accountdb/balance
// Shows mob's ID's current funds as a big overlay, static over a mob if under 150
/datum/visioneffect/antag
// Shows ALL factions on a mob
/datum/visioneffect/construct
// Shows construct health
/datum/visioneffect/cult_conversion
// Shows cult convertability status
/datum/visioneffect/diagnostic
// Shows health and energy levels of borgs and mechs
/datum/visioneffect/implant
// Shows implants a target has, including loyalty, tracking, etc
/datum/visioneffect/job
// Shows what job a mob's ID has
/datum/visioneffect/material
// Shows the original pipe layout of a station
/datum/visioneffect/medical
// Shows medical HUD information, allows full medical record access on examine
/datum/visioneffect/meson
// Mesons!
/datum/visioneffect/night
// Shows a green overlay and operates exactly how you think night vision should
/datum/visioneffect/nullrod
// Shows a pink icon next to a mob if they are carrying the null rod
/datum/visioneffect/pathogen
// Shows all the pathogen stuff with a purple overlay
/datum/visioneffect/security
// Shows the ARREST status of a person (not job or implants!) and gives basic security record stauts on examine. Does not let you set arrest status, just view
/datum/visioneffect/security/arrest
// As above, but this version lets you set people to arrest
/datum/visioneffect/thermal
// Shows mobs through walls!
/datum/visioneffect/vampire_improved
// Gives thermal vision, the 'your vision has improved' vampire effect
/datum/visioneffect/vampire_mature
// Fully powered vampire. Can see everything
```

## Adminbus Toys
image
This PR adds a few adminbus HUDs for icon conflict reasons. Later PRs will determine if they're coming to a station near you, or if they'll remain a funny bus toy.

image
Combined Med/Sec HUDs, which does what it says on the tin... but it can't set people to arrest. Can be made by combining a security scouter with the medical hud. It can also be split apart. (Older image showing possible future medhud icon use, not included in this PR.)

image
error
dogwalk
Wage HUDs, see people's job, wage, and employment history!
Cash HUDs. If you don't have at least 150 in your bank account, you aren't a person. (gif above from an earlier version that needed like 400)

Golden Aviators, they're golden sunglasses that show your job.

Bonus image from testing:
image

This PR has been tested both with and without Europa Lights, and it works just fine in both settings.

## Why it's good
d a t u m s

This will let us do things like put meson vision into cybereyes without having to hardcode mesons again. It'll also let us stack these huds, so you can have medical and sechuds on at the same time. Imagine a sword that gives you thermal vision when you have it on your person. Or an admin decides to give you permanent sechuds in a TC trade. Things like that.

## SPECIAL THANKS
kilozombie for the equipped versions of the double cursed huds
Blithering for the bunches of sunglasses sprites

## Changelog

:cl:
 * rscadd: Updated HUDs to use Datums. Please bug report any issues with sechuds, mesons, any sort of vision-enchanting set of glasses.
 * tweak: Security/Medical HUDs examining people has been formatted a little better.
 * bugfix: Fixed a bug where skeletons can completely lose any ability to see in the dark.
 * imageadd: New sunglasses and other icons added to the code for adminbus and future use.

[bugfix] [sprites] [system] [tested] [ui]

### CODE REVIEW

**Comment by Eneocho** [comment]
Does this fix the detective's Investigation Glasses? The PR said (and the code implies) that they have both the tracking glasses features and the sechud, but last I tested them the sechud was broken and didn't appear. Does this revamp that so it actually works?

**Comment by jwhitak** [comment]
It does not. Let me fix that real fast.

**Comment by jwhitak** [comment]
image
Caught another miss while fixing this up. Investigation Glasses should now be good to go.

**Comment by GreatBigFailure** [comment]
Oh hey they're here. Looks good, wage hud is 10/10 
Looking forward to when they're naturally obtainable in game, glad it's far enough along to merge a bus version though

**Comment by Eneocho** [comment]
There could probably be an argument for a "walletsight" HUD that lets you see the current moneys someone has, as a trader tool.

**Comment by jwhitak** [comment]
> There could probably be an argument for a "walletsight" HUD that lets you see the current moneys someone has, as a trader tool.

If this gets merged, you can look at the visioneffect/account huds to get an idea on how to make your own!

**Comment by DeityLink** [comment]
It's finally here!

File: code/datums/hud/account_hud.dm [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+				cashdisplay = html_encode(Gibberish("ERRO",100))
+			else if (HUD_ACCOUNT_DB_OFFLINE) // Have to call this after the first if because it will runtime if linked_db doesn't exist
+				cashdisplay = html_encode(Gibberish("ERRO",100))
+			else //not an error case
+				if(virtualacc)
+					cash += virtualacc.money
+				if(bankacc)
+					cash += bankacc.money
+				cashdisplay = "$[cash]"
+			if(cash < 150)
```
Comment (Kurfursten) [inline code/datums/hud/account_hud.dm]: turn magic number into a define

**Review by Kurfursten — APPROVED** [review]
Codewise this seems fine, it was wise to not make obtainable the new glasses (syndishades, brig medic scouter, aviators, etc.) until a later PR.

Fix that one odd magic number

**Comment by Kurfursten** [comment]
It would be nice to see some other enhanced glasses in a later PR if you have an interest in more features down this route. Here are a few of my ideas --

> CE

Option to toggle on meson or material scanner mode as a simple option.

There's an idea I've had for ages that might be way too trouble to code where it would show the wattage in a grid if you can see an exposed wire (only on the nearest wire within that grid). This basically saves you a multitool scan but would also make it way easier to locate a break in a powerline due to floating numbers.

> RD

Combines superior welding goggle protection (spawns in RD office) with mesons (for xenoarch work) and diagnostic HUD (for mechs and cyborgs) - a very respectable suite I think.

> HydroHUD

Scrapped features for this were going to have it display overlays on animals e.g. chicken feed level (governs eggs), cow/goat milk level, boxen growth progress, gourmonger hunger, etc. and show the tray warning lights on normal soil plots. I ended up not coding this due to exhaustion with the crate but if you love HUDs this is definitely a feature worth revisiting.

File: code/datums/hud/account_hud.dm (hunk 1 above)
Comment (jwhitak) [reply code/datums/hud/account_hud.dm]: I'll turn it into a variable inside of the datum itself, the best solution. Later glasses might let you configure this line, or change it under conditions. Admins can also bus changes into it (edit this variable in the instance found in the stored_hud list of the balance viewing glasses)

**Pull request merged**